test/generate/test_generate.py

```diff
@@ -0,0 +1,256 @@
+
+
+def example_generate(
+    config_path: str,
+    checkpoint_path: str,
+    prompt: str,
+    *,
+    device: str = "cuda",
+    temperature: float = 1.0,
+    max_new_tokens: int = 32,
+    batch_size: int = 1,
+    top_k: Optional[int] = None,
+    seed: Optional[int] = None,
+):
+    init_logger()
+    color = utils.Color
+
+    # Load configuration from toml file
+    config = JobConfig()
+    config.parse_args([f"--job.config_file={config_path}"])
+    config._validate_config()
+
+    utils.set_determinism(seed)
+
+    world_size = int(os.environ.get("WORLD_SIZE", 1))
+    local_rank = int(os.environ.get("LOCAL_RANK", 0))
+    device = torch.device(f"cuda:{local_rank}")
+    torch.cuda.set_device(device)
+    gpu_memory_monitor = build_gpu_memory_monitor()
+
+    model_name = config.model.name
+
+    # Init distributed env
+    if world_size > 1:
+        utils.init_distributed(config)
+        parallel_dims = ParallelDims(
+            dp_replicate=1,
+            dp_shard=-1,
+            cp=1,
+            tp=world_size,
+            pp=1,
+            world_size=world_size,
+            enable_loss_parallel=False,
+        )
+        # Build world mesh for parallelism
+        world_mesh = parallel_dims.build_mesh(device_type="cuda")
+
+    logger.info(f"World Size: {world_size}, Local Rank: {local_rank} on {device}")
+
+    # Tokenizer setup
+    tokenizer = build_tokenizer(
+        model_name_to_tokenizer[model_name], config.model.tokenizer_path
+    )
+
+    model_config = models_config[model_name][config.model.flavor]
+    model_config.norm_type = config.model.norm_type
+    model_config.max_seq_len = config.training.seq_len
+    model_config.vocab_size = tokenizer.n_words
+
+    model_cls = model_name_to_cls[model_name]
+    init_device = "meta" if world_size > 1 else device
+    with torch.device(init_device):
+        logger.info(f"Init model on init_device: {init_device}")
+        model = model_cls.from_model_args(model_config)
+
+    if world_size > 1:
+        models_parallelize_fns[model_name](model, world_mesh, parallel_dims, config)
+
+    # materalize model
+    model.to_empty(device="cuda")
+    model.eval()
+
+    state_dict = {"model": model.state_dict()}
+
+    # Checkpoint Loading
+    begin = time.monotonic()
+    logger.info(f"Loading chkpt at: {checkpoint_path}")
+    dcp.load(state_dict, checkpoint_id=checkpoint_path)
+    logger.info(f"Finished loading chkpt in {time.monotonic() - begin:.2f} seconds.")
+
+    gpu_mem_stats = gpu_memory_monitor.get_peak_stats()
+    logger.info(
+        f"GPU memory usage for model: "
+        f"{gpu_mem_stats.max_reserved_gib:.2f}GiB"
+        f"({gpu_mem_stats.max_reserved_pct:.2f}%)"
+    )
+
+    # Tokenize prompt and repeat batch_size times
+    input_ids = (
+        (
+            torch.tensor(
+                tokenizer.encode(prompt, bos=True, eos=False), dtype=torch.long
+            )
+            .view(1, -1)
+            .repeat(batch_size, 1)
```

Comment:
Question: Why we would run `batch_size` number of same inference tasks? I thought `batch_size` can be useful when we have multiple distinct prompts. In our case maybe it's not necessary?